packages/ember-htmlbars/lib/hooks/create-fresh-scope.js

```diff
@@ -46,13 +49,124 @@
   the current view's `controller`.
 */
 
+function Scope(parent) {
+  this._self = null;
+  this._blocks = null;
+  this._component = null;
+  this._view = null;
+  this._attrs = null;
+  this._locals = null;
+  this._localPresent = null;
+  this.overrideController = null;
+  this.parent = parent;
+}
+
+let proto = Scope.prototype;
+
+proto.getSelf = function() {
+  return this._self || this.parent.getSelf();
+};
+
+proto.bindSelf = function(self) {
+  this._self = self;
+};
+
+proto.updateSelf = function(self, key) {
+  let existing = this._self;
+
+  if (existing) {
+    existing.setSource(self);
+  } else {
+    this._self = new ProxyStream(self, key);
+  }
+};
+
+proto.getBlock = function(name) {
```

Comment:
lots of functions here, they can likely be extracted into shared macros (to reduce bytes, without causing perf pains)